### PR TITLE
Resolve thread sanitizer issues

### LINKED ISSUE #5031: Data races on startup
**To Reproduce**
Steps to reproduce the behavior from a clean checkout:
```bash
mkdir build && cd build/
CC=clang-10 CXX=clang++-10 cmake -DCMAKE_CXX_FLAGS="-fsanitize=thread" -DCMAKE_EXE_LINKER_FLAGS="-fsanitize=thread" -DOPTION_ASAN=OFF ..
make
cd ..
./build/src/widelands --datadir=./data/
```

**Expected behavior**
No data race

**Actual behavior**
```
==================
WARNING: ThreadSanitizer: data race (pid=57617)
  Write of size 8 at 0x000002c22078 by main thread:
    #0 UI::Panel::ModalGuard::ModalGuard(UI::Panel&) /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:172:16 (widelands+0xc775e0)
    #1 UI::Panel::do_run() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:356:13 (widelands+0xc79b4b)
    #2 int UI::Panel::run<int>() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.h:120:34 (widelands+0x6c6dd8)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:813:5 (widelands+0x6b76ed)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

  Previous read of size 8 at 0x000002c22078 by thread T3:
    #0 UI::Panel::logic_thread() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:192:6 (widelands+0xc7774a)
    #1 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14 (widelands+0x6d3e1d)
    #2 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14 (widelands+0x6d3d20)
    #3 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13 (widelands+0x6d3cc8)
    #4 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11 (widelands+0x6d3c68)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13 (widelands+0x6d3b2f)
    #6 <null> <null> (libstdc++.so.6+0xd6de3)

  Location is global 'UI::Panel::modal_' of size 8 at 0x000002c22078 (widelands+0x000002c22078)

  Thread T3 (tid=57628, running) created by main thread at:
    #0 pthread_create <null> (widelands+0x61b2eb)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd70a8)
    #2 (anonymous namespace)::GameLogicThread::GameLogicThread(bool*) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:226:32 (widelands+0x6b79db)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:744:18 (widelands+0x6b5669)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

SUMMARY: ThreadSanitizer: data race /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:172:16 in UI::Panel::ModalGuard::ModalGuard(UI::Panel&)
==================
==================
WARNING: ThreadSanitizer: data race (pid=57617)
  Read of size 4 at 0x7fff11f7abf0 by thread T3:
    #0 UI::Panel::logic_thread() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:194:16 (widelands+0xc7777b)
    #1 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14 (widelands+0x6d3e1d)
    #2 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14 (widelands+0x6d3d20)
    #3 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13 (widelands+0x6d3cc8)
    #4 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11 (widelands+0x6d3c68)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13 (widelands+0x6d3b2f)
    #6 <null> <null> (libstdc++.so.6+0xd6de3)

  Previous write of size 4 at 0x7fff11f7abf0 by main thread:
    [failed to restore the stack]

  As if synchronized via sleep:
    #0 nanosleep <null> (widelands+0x618b3d)
    #1 SDL_Delay_REAL <null> (libSDL2-2.0.so.0+0x14014d)
    #2 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14 (widelands+0x6d3e1d)
    #3 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14 (widelands+0x6d3d20)
    #4 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13 (widelands+0x6d3cc8)
    #5 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11 (widelands+0x6d3c68)
    #6 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13 (widelands+0x6d3b2f)
    #7 <null> <null> (libstdc++.so.6+0xd6de3)

  Location is stack of main thread.

  Location is global '??' at 0x7fff11f63000 ([stack]+0x000000017bf0)

  Thread T3 (tid=57628, running) created by main thread at:
    #0 pthread_create <null> (widelands+0x61b2eb)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd70a8)
    #2 (anonymous namespace)::GameLogicThread::GameLogicThread(bool*) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:226:32 (widelands+0x6b79db)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:744:18 (widelands+0x6b5669)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

SUMMARY: ThreadSanitizer: data race /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:194:16 in UI::Panel::logic_thread()
==================
==================
WARNING: ThreadSanitizer: data race (pid=57617)
  Write of size 1 at 0x7b44000001cc by main thread:
    #0 WLApplication::handle_input(InputCallback const*) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:984:16 (widelands+0x6b8f0a)
    #1 UI::Panel::do_run() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:399:9 (widelands+0xc79f9a)
    #2 int UI::Panel::run<int>() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.h:120:34 (widelands+0x6c6dd8)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:813:5 (widelands+0x6b76ed)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

  Previous read of size 1 at 0x7b44000001cc by thread T3:
    #0 WLApplication::should_die() const /home/ceeac/Projects/code/widelands/src/wlapplication.h:146:10 (widelands+0xc83ec9)
    #1 UI::Panel::logic_thread() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:190:15 (widelands+0xc77729)
    #2 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14 (widelands+0x6d3e1d)
    #3 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14 (widelands+0x6d3d20)
    #4 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13 (widelands+0x6d3cc8)
    #5 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11 (widelands+0x6d3c68)
    #6 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13 (widelands+0x6d3b2f)
    #7 <null> <null> (libstdc++.so.6+0xd6de3)

  As if synchronized via sleep:
    #0 nanosleep <null> (widelands+0x618b3d)
    #1 SDL_Delay_REAL <null> (libSDL2-2.0.so.0+0x14014d)
    #2 int UI::Panel::run<int>() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.h:120:34 (widelands+0x6c6dd8)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:813:5 (widelands+0x6b76ed)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

  Location is heap block of size 320 at 0x7b4400000140 allocated by main thread:
    #0 operator new(unsigned long) <null> (widelands+0x6a81fb)
    #1 WLApplication::get(int, char const**) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:329:19 (widelands+0x6abc43)
    #2 main /home/ceeac/Projects/code/widelands/src/main.cc:45:11 (widelands+0x6aa85f)

  Thread T3 (tid=57628, running) created by main thread at:
    #0 pthread_create <null> (widelands+0x61b2eb)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd70a8)
    #2 (anonymous namespace)::GameLogicThread::GameLogicThread(bool*) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:226:32 (widelands+0x6b79db)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:744:18 (widelands+0x6b5669)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

SUMMARY: ThreadSanitizer: data race /home/ceeac/Projects/code/widelands/src/wlapplication.cc:984:16 in WLApplication::handle_input(InputCallback const*)
==================
==================
WARNING: ThreadSanitizer: data race (pid=57617)
  Read of size 1 at 0x000002c220b0 by main thread:
    #0 UI::Panel::do_run() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:433:68 (widelands+0xc7a2cd)
    #1 int UI::Panel::run<int>() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.h:120:34 (widelands+0x6c6dd8)
    #2 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:813:5 (widelands+0x6b76ed)
    #3 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

  Previous write of size 1 at 0x000002c220b0 by thread T3:
    #0 UI::Panel::logic_thread() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:187:24 (widelands+0xc776fd)
    #1 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:60:14 (widelands+0x6d3e1d)
    #2 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/bits/invoke.h:95:14 (widelands+0x6d3d20)
    #3 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:264:13 (widelands+0x6d3cc8)
    #4 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:271:11 (widelands+0x6d3c68)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/10/../../../../include/c++/10/thread:215:13 (widelands+0x6d3b2f)
    #6 <null> <null> (libstdc++.so.6+0xd6de3)

  As if synchronized via sleep:
    #0 nanosleep <null> (widelands+0x618b3d)
    #1 SDL_Delay_REAL <null> (libSDL2-2.0.so.0+0x14014d)
    #2 int UI::Panel::run<int>() /home/ceeac/Projects/code/widelands/src/ui_basic/panel.h:120:34 (widelands+0x6c6dd8)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:813:5 (widelands+0x6b76ed)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

  Location is global 'UI::Panel::logic_thread_running_' of size 1 at 0x000002c220b0 (widelands+0x000002c220b0)

  Thread T3 (tid=57628, running) created by main thread at:
    #0 pthread_create <null> (widelands+0x61b2eb)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd70a8)
    #2 (anonymous namespace)::GameLogicThread::GameLogicThread(bool*) /home/ceeac/Projects/code/widelands/src/wlapplication.cc:226:32 (widelands+0x6b79db)
    #3 WLApplication::run() /home/ceeac/Projects/code/widelands/src/wlapplication.cc:744:18 (widelands+0x6b5669)
    #4 main /home/ceeac/Projects/code/widelands/src/main.cc:47:10 (widelands+0x6aa87f)

SUMMARY: ThreadSanitizer: data race /home/ceeac/Projects/code/widelands/src/ui_basic/panel.cc:433:68 in UI::Panel::do_run()
==================
```
**Version:**
 - OS: Ubuntu 20.04
 - Widelands Version: `40dc16d84eaacce97cb085ea126f63d15a6f6841` (latest master branch as of today)

### PR DESCRIPTION
**Pull request by matthiakl**
- Added option to compile with thread sanitizer for future multithreading debugging
- Fixed #5031 (apart from some 3rd party issues in libpulse) 
- Reduced issues reported during creation of a new game from 87 to 33, mostly by using atomic types.

The remaining issues can be boiled down to the `NoteThreadSafeFunction` handling, which I currently don't quite understand ;)
[tsan.txt](https://github.com/widelands/widelands/files/7532304/tsan.txt)

### CODE REVIEW

File: src/map_io/map_players_view_packet.cc [hunk 1]
```diff
@@ -385,8 +385,8 @@ void MapPlayersViewPacket::read(FileSystem& fs, EditorGameBase& egbase) {
 					f.resource_amounts.d = fr.unsigned_8();
 					f.resource_amounts.r = fr.unsigned_8();
 
-					f.terrains.d = fr.unsigned_32();
-					f.terrains.r = fr.unsigned_32();
+					f.terrains.store({static_cast<DescriptionIndex>(fr.unsigned_32()),
+					                  static_cast<DescriptionIndex>(fr.unsigned_32())});
```
Comment (Noordfrees) [inline src/map_io/map_players_view_packet.cc]: Calling several function of FileRead in the same statement is dangerous, please store them in local variables first. 
Guaranteed execution order? Some compilers are above that ;)

File: src/logic/map_objects/tribes/soldier.cc [hunk 2]
```diff
@@ -1272,7 +1272,7 @@ void Soldier::defense_update(Game& game, State& state) {
 		// If enemy is in our land, then go after it!
 		if (upcast(Soldier, soldier, temp_bob)) {
 			assert(soldier != this);
-			Field const f = game.map().operator[](soldier->get_position());
+			const Field& f = game.map().operator[](soldier->get_position());
```
Comment (Noordfrees) [inline src/logic/map_objects/tribes/soldier.cc:1275]: Good catch :+1:

**Review by Noordfrees — COMMENTED** [review]
Looks very good, will try this out later :)

**Comment by klaus-halfmann** [comment]
Checking this, too. Maybe I find my  #5111

**Review by klaus-halfmann — CHANGES_REQUESTED** [review]
Did not compile:
```
5031-tsan/src/logic/game.h:42:20: error: constexpr variable 'kStatisticsSampleTime' must be initialized by a constant expression
constexpr Duration kStatisticsSampleTime = Duration(30 * 1000);
                   ^                       ~~~~~~~~~~~~~~~~~~~
031-tsan/src/logic/game.h:42:44: note: non-constexpr constructor 'Duration' cannot be used in a constant expression
constexpr Duration kStatisticsSampleTime = Duration(30 * 1000);
                                           ^
5031-tsan/src/base/times.h:38:2: note: declared here
        Duration(const Duration& other) : value_(other.get()) {
        ^
1 error generated.
```

**Comment by klaus-halfmann** [comment]
@matthiakl I still get the compiler error and clang_tidy is failing shall I try to fix these?

**Comment by matthiakl** [comment]
> @matthiakl I still get the compiler error and clang_tidy is failing shall I try to fix these?

I am working on it, gcc does too lose checks for constexpr, so I have to check with clang.

**Comment by matthiakl** [comment]
@klaus-halfmann it should compile now with other compilers.

**Comment by klaus-halfmann** [comment]
Thanks, compiles now, but I still get #5111 without further clue. Will do codecheck later.

**Comment by klaus-halfmann** [comment]
Uhm, how Do I activate the TSAN?

**Comment by klaus-halfmann** [comment]
Those docs should help https://www.cplusplus.com/reference/atomic/atomic/

**Review by klaus-halfmann — COMMENTED** [review]
Found lots of improved CTor calls, fine.
Found usage of atomic<T> and load/sroe operations, fine.

I am still missing the big picture, Can you point to some background article?
Can we have a Jitsi call where you explain how you are doing this?
What _must_ be tested before this can be merged?

**Comment by matthiakl** [comment]
> Uhm, how Do I activate the TSAN?

You need  to run cmake with  `-DOPTION_ASAN=OFF -DOPTION_TSAN=ON` 

Basically I looked which variables were accessed from multiple threads as reported by tsan. Then I tried to protect the access by making it atomic.

**Comment by hessenfarmer** [comment]
@matthiakl  although I really don't understand what is necessary neither what you did here I got the impression this might help with the task of https://github.com/widelands/widelands/pull/5013.
I really got stuck there. maybe you might have a look once this had been merged

**Comment by klaus-halfmann** [comment]
OK, trying `/compile.sh --debug --no-asan -DOPTION_TSAN=ON `

Do ASAN and TSAN not work together or would it be just to slow?

https://clang.llvm.org/docs/ThreadSanitizer.html tells me 
 
> Typical slowdown introduced by ThreadSanitizer is about 5x-15x. Typical memory overhead introduced by ThreadSanitizer is about 5x-10x.

this will be rather slow, OK

**Review by klaus-halfmann — APPROVED** [review]
OK, this _is_ really slow. But I got no TSAN Warnings on startup,
Continuing a game gave me however lots of them, where can we collect these?

**Comment by matthiakl** [comment]
> I really got stuck there. maybe you might have a look once this had been merged

I will take a look into it and see if I can help.

> Do ASAN and TSAN not work together or would it be just to slow?

No, they cannot be enabled together (at least on gcc) and yes they both slow down execution considerably.

> OK, this is really slow. But I got no TSAN Warnings on startup,

That's good. On master you will get some warnings. Since you are on Mac, I suspect you have another audio library than libpulse which created some warnings on my system.

> Continuing a game gave me however lots of them, where can we collect these?

We should open a new issue, to resolve remaining issues in a future branch.

**Review by Noordfrees — COMMENTED** [review]
Please use spaces instead of tabs for indentation in CMake files for consistency. And the compile.sh short option should be kept to 1 character, it doesn't need to be related to the long name if we're running out of letters.

**Comment by matthiakl** [comment]
> Please use spaces instead of tabs for indentation in CMake files for consistency

Stupid editor settings…
I chose `m` now as short option (multithreading) for thread sanitizer.
The linkage with atomic is a bit weird: clang on linux needs it explicitly while on mac it is builtin. It should compile now on all platforms.

**Comment by klaus-halfmann** [comment]
Used new compile script, fine.
Still OK for me. created the expectable followup Issues: #5119 #5120

**Review by Noordfrees — APPROVED** [review]

**Pull request merged**